charon/src/bin/charon-driver/translate/translate_types.rs

```diff
@@ -424,6 +429,55 @@ impl<'tcx, 'ctx> ItemTransCtx<'tcx, 'ctx> {
             }
         }
 
+        fn translate_variant_id(r_id: r_abi::VariantIdx) -> VariantId {
+            VariantId::from_usize(r_abi::VariantIdx::as_usize(r_id))
+        }
+
+        // Computes the tag representation of the variant's discriminant if possible.
+        //
+        // If the discriminant is encoded as a niche the following holds:
+        // If discriminant != self.discriminant_layout.untagged_variant
+        // then tag = (d-self.discriminant_layout.tagged_variants_start).wrapping_add(self.discriminant_layout.niche_start)
+        //
+        // Note: it is possible that the tag is stored in the niche of a pointer type, but will
+        // be returned as an integer instead. This is supposed to be a different interpretation of the same bytes.
+        fn translate_discr_to_tag(
+            discr: ScalarValue,
+            tag_ty: IntegerTy,
+            variants: &Vector<VariantId, Variant>,
+            encoding: &r_abi::TagEncoding<r_abi::VariantIdx>,
+        ) -> Option<ScalarValue> {
+            match &encoding {
+                // The direct encoding is just a cast.
+                r_abi::TagEncoding::Direct => Some(ScalarValue::from_bits(tag_ty, discr.to_bits())),
+                r_abi::TagEncoding::Niche {
+                    untagged_variant,
+                    niche_variants,
+                    niche_start,
+                } => {
+                    let get_bits_of_variant = |variant: r_abi::VariantIdx| {
+                        let variant = translate_variant_id(variant);
+                        variants
+                            .get(variant)
+                            .expect("Variant index out of bounds while translating tag")
+                            .discriminant
+                            .to_bits()
+                    };
+                    let discr_bits = discr.to_bits();
+                    if discr_bits == get_bits_of_variant(*untagged_variant) {
+                        None // This variant does not have a tag.
+                    } else {
+                        let discr_rel = discr_bits - get_bits_of_variant(*niche_variants.start());
```

Comment:
That's not [the formula used by cg_clif](https://github.com/rust-lang/rustc_codegen_cranelift/blob/8dea73acfd76a38173975ff1e60b7770af677768/src/discriminant.rs#L58-L70). Discriminants don't play into it; instead, it's only comparisons on variant ids. This function should take the current variant_id as input, then
```rust
let discr_rel = variant_index.as_u32() - niche_variants.start().as_u32();
let tag_bits = (discr_rel as u128).wrapping_add(niche_start);
```